tests/ttnn/unit_tests/gtests/ccl/kernels/fabric_worker_sender_multi_input.cpp

```diff
@@ -0,0 +1,222 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "dataflow_api.h"
+#include "ttnn/cpp/ttnn/operations/ccl/kernels/edm_fabric/fabric_edm_packet_header.hpp"
+#include "ttnn/cpp/ttnn/operations/ccl/kernels/edm_fabric/edm_fabric_worker_adapters.hpp"
+#include "tests/ttnn/unit_tests/gtests/ccl/kernels/test_kernels.common.hpp"
+
+struct unicast_mode {
+    uint8_t distance;
+};
+struct mcast_mode {
+    uint8_t distance;
+    uint8_t range;
+};
+
+union transmit_config {
+    unicast_mode unicast;
+    mcast_mode mcast;
+};
+
+enum class ReadMode {
+    // Fully drain one input CB before advancing to the next
+    FULLY_ORDERED,
+    // Drain both inputs in at some specified ratio (e.g. if 5:3 then 3 pages from CB0 then 5 from CB1)
+    RATIOD_FORWARDING,
+    // Read pages from either CB as soon as they are available
+    ARBITRARILY_ORDERED
+};
+
+static constexpr size_t NORMALIZED_NOC_INDEX = 0;
+
+template <typename AddrGen>
+auto forward_to_fabric_from_cb(
+    size_t total_pages_to_send,
+    tt::fabric::WorkerToFabricEdmSender &sender,
+    uint32_t cb_id,
+    transmit_config const& config,
+    size_t page_size,
+    const AddrGen &dest_addr_gen,
+    size_t num_pages_per_send,
+    size_t current_page
+    ) {
+
+    // for (uint32_t p = 0; p < total_pages_to_send; p += num_pages_per_send) {
+    uint32_t pages_to_send = std::min<uint32_t>(num_pages_per_send, total_pages_to_send - current_page);
+    sender.wait_for_empty_write_slot();
+
+    // bit of a hack to extract X/Y
+    const auto dest_noc_address = get_noc_addr(current_page, dest_addr_gen, 0, NORMALIZED_NOC_INDEX);
+    const size_t dest_addr = dest_noc_address & 0xFFFFFFFF;
+    const size_t dest_noc_x = (dest_noc_address >> NOC_ADDR_LOCAL_BITS) & ((1 << NOC_ADDR_NODE_ID_BITS) - 1);
+    const size_t dest_noc_y = (dest_noc_address >> (NOC_ADDR_LOCAL_BITS + NOC_ADDR_NODE_ID_BITS)) & ((1 << NOC_ADDR_NODE_ID_BITS) - 1);
+    const size_t packet_size = page_size + sizeof(tt::fabric::PacketHeader);
+
+    auto packet_addr = get_read_ptr(cb_id);
+    auto &packet_header = *reinterpret_cast<tt::fabric::PacketHeader*>(packet_addr);
+    if constexpr (mcast_mode) {
+        packet_header.to_write()
+            .to_chip_multicast(tt::fabric::MulticastRoutingCommandHeader{config.mcast.distance, config.mcast.range})
+            .to_noc_unicast(tt::fabric::NocUnicastCommandHeader{
+                dest_addr,
+                (pages_to_send * page_size) + sizeof(tt::fabric::PacketHeader),
+                static_cast<uint8_t>(dest_noc_x),
+                static_cast<uint8_t>(dest_noc_y)
+            });
+        packet_header.reserved2 = 0x1111; // debug only
```

Comment:
An explanation is needed for the 1111 magic number

Reply:
constexpr uint32_t kDebugOnlyTagForPacket = 0x1111 :)